### PR TITLE
Support `--workdir-root` in the `tmt clean` command

### PR DESCRIPTION
**Pull request by skycastlelily**
Pull Request Checklist

* [x] implement the feature
* [x] extend the test coverage
* [x] include a release note

### CODE REVIEW

File: tmt/base.py [hunk 1]
```diff
@@ -67,11 +67,12 @@
     cached_property,
     container_field,
     dict_to_yaml,
+    effective_workdir_root,
     field,
     git_clone,
     normalize_shell_script,
     verdict,
-    )
+)
```
Comment (happz) [inline tmt/base.py:75]: Two unrelated formatting changes, here and on line 26.

Comment (skycastlelily) [reply tmt/base.py]: updated^^

**Comment by skycastlelily** [comment]
Hi Petr, @psss, would you please spare some time to review this merge request,please? This is supposed to address #2807 , sorry  for forgetting mentioned it before 1.36 release,and hope we can have this merged for 1.37: )

File: docs/releases.rst [hunk 2]
```diff
@@ -7,6 +7,9 @@
 tmt-1.37.0
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+Add support --workdir-root for clean function, users can specify
```
Comment (thrix) [inline docs/releases.rst:10]:
```suggestion
Add support ``--workdir-root`` for clean function. Users can specify
```

File: docs/releases.rst [hunk 3]
```diff
@@ -7,6 +7,9 @@
...
+the dir they want to clean things for.
```
Comment (thrix) [inline docs/releases.rst:11]:
```suggestion
the directory they want to clean things for.
```

File: tmt/cli.py [hunk 4]
```diff
@@ -1907,6 +1909,9 @@ def clean(context: Context,
         raise tmt.utils.GeneralError(
             "Options --last and --id cannot be used together.")
 
+    if workdir_root and not Path(workdir_root).exists():
+        raise tmt.utils.GeneralError(f"Path '{workdir_root}' doesn't exist.")
```
Comment (martinhoyer) [inline tmt/cli.py:1913]: Just checking - is it expected to error out here? I'm thinking if user want to make sure some path that doesn't exist is "cleaned", a warning would suffice?

File: tmt/utils/__init__.py [hunk 5]
```diff
@@ -230,13 +230,16 @@ def format(cls, exit_code: int) -> Optional[str]:
...
-def effective_workdir_root() -> Path:
+def effective_workdir_root(workdir_root_option: Optional[str] = None) -> Path:
     """
     Find out what the actual workdir root is.
 
-    If ``TMT_WORKDIR_ROOT`` variable is set, it is used as the workdir root.
-    Otherwise, the default of :py:data:`WORKDIR_ROOT` is used.
+    If ``workdir-root`` cli option is set, it is used as the workdir root.
+    Otherwise, ``TMT_WORKDIR_ROOT`` variable is used, if it is set. Or, the default
+    of :py:data:`WORKDIR_ROOT` is used.
```
Comment (martinhoyer) [inline tmt/utils/__init__.py:239]:
```suggestion
    If the `workdir-root` cli option is set, it is used as the workdir root.
    Otherwise, the ``TMT_WORKDIR_ROOT`` environment variable is used if set.
    If neither is specified, the default value of :py:data:`WORKDIR_ROOT` is used.
```

**Review by martinhoyer — COMMENTED** [review]
lgtm. Just nitpicks.  
I'm surprised setting a workdir root is so complicated tbh (not necessarily related to these changes). :)

**Comment by thrix** [comment]
@skycastlelily would it be possible to log the workdir root please in case the verbose mode is enabled please?

Something like:

```
❯ env | grep WORKDIR
TMT_WORKDIR_ROOT=/var/home/mvadkert/.local/share/tmt

❯ tmt clean -v
clean
    workdir /var/home/mvadkert/.local/share/tmt
guests
runs
images
    testcloud

```

File: tmt/utils/__init__.py (hunk 5 above)
Comment (thrix) [reply tmt/utils/__init__.py:239]: @martinhoyer why a single backtick btw?

File: tmt/base.py [hunk 6]
```diff
@@ -3900,7 +3902,7 @@ def _stop_running_guests(self, run: Run) -> bool:
     def guests(self, run_ids: tuple[str, ...]) -> bool:
         """ Clean guests of runs """
         self.info('guests', color='blue')
-        root_path = Path(self.opt('workdir-root'))
+        root_path = Path(self.workdir_root)
```
Comment (happz) [inline tmt/base.py:3905]: `Path()` may not be needed because `self.workdir_root` is already returning `Path` instances.

File: tmt/base.py [hunk 7]
```diff
@@ -3934,7 +3936,7 @@ def _clean_workdir(self, path: Path) -> bool:
     def runs(self, id_: tuple[str, ...]) -> bool:
         """ Clean workdirs of runs """
         self.info('runs', color='blue')
-        root_path = Path(self.opt('workdir-root'))
+        root_path = Path(self.workdir_root)
```
Comment (happz) [inline tmt/base.py:3939]: `Path()` may not be needed because `self.workdir_root` is already returning `Path` instances.

File: tmt/base.py (hunk 6 above)
Comment (happz) [reply tmt/base.py:3905]: And since this variable is used exactly once, on like 3913, we can easily not create it and use `self.workdir_root` when calling `generate_runs()`.

File: tmt/base.py (hunk 7 above)
Comment (happz) [reply tmt/base.py:3939]: Same as above, `Path()` is not needed and variable is used just once, we can pass `self.workdir_root` to `generate_runs()`.

File: tmt/utils/__init__.py [hunk 8]
```diff
@@ -230,13 +230,16 @@ def format(cls, exit_code: int) -> Optional[str]:
 T = TypeVar('T')
 
 
-def effective_workdir_root() -> Path:
+def effective_workdir_root(workdir_root_option: Optional[str] = None) -> Path:
```
Comment (happz) [inline tmt/utils/__init__.py:233]: I'd go with `Optional[Path]` already: as soon as we get the value of `--workdir-root` option in CLI commands, we should convert it to either a `None` or `Path` instance.

File: tmt/options.py [hunk 9]
```diff
@@ -183,12 +183,8 @@ def option(
...
 WORKDIR_ROOT_OPTIONS: list[ClickOptionDecoratorType] = [
     option(
-        '--workdir-root', metavar='PATH', envvar='TMT_WORKDIR_ROOT',
-        default=tmt.utils.WORKDIR_ROOT,
-        help=f"""
-             Path to root directory containing run workdirs.
-             Defaults to '{tmt.utils.WORKDIR_ROOT}'.
-             """)
+        '--workdir-root', metavar='PATH',
+        help='Path to root directory containing run workdirs.'),
```
Comment (happz) [inline tmt/options.py:187]: I'd argue the default value and envvar should be still used, `TMT_WORKDIR_ROOT` would still be considered as "workdir root", and so on. Is there a reason to remove it?

File: tmt/base.py (hunk 6 above)
Comment (skycastlelily) [reply tmt/base.py]: Yeah, but  after this update, root_path will be called twice, we'd better create it,right?:)

File: tmt/cli.py (hunk 4 above)
Comment (skycastlelily) [reply tmt/cli.py:1913]: yeah,it should be suffice in that case, but in that case,it won't hurt,well,so much,if we give an error.And the error would be for other cases, if we don't error out, tmt will continue the following steps,which will do nothing for users,but just waste their time,right?:)

File: tmt/options.py (hunk 9 above)
Comment (skycastlelily) [reply tmt/options.py:190]: I think we have effective_workdir_root to take care of the default now, but sure, we can also add the default value here:)

**Comment by skycastlelily** [comment]
Updated^^

File: tmt/base.py [hunk 10]
```diff
@@ -3900,7 +3902,8 @@ def _stop_running_guests(self, run: Run) -> bool:
     def guests(self, run_ids: tuple[str, ...]) -> bool:
         """ Clean guests of runs """
         self.info('guests', color='blue')
-        root_path = Path(self.opt('workdir-root'))
+        root_path = self.workdir_root
+        self.verbose(f"workdir: {root_path}")
```
Comment (happz) [inline tmt/base.py:3906]: 1. it's not a workdir you're logging, it's a workdir root,
2. `verbose()` and other logging methods can handle label/value or key/value pairs out of the box, taking care of indentation and colon. Together, the logging should look like this: `self.verbose('workdir', root_path)`

File: tmt/utils/__init__.py [hunk 11]
```diff
@@ -230,13 +230,16 @@ def format(cls, exit_code: int) -> Optional[str]:
...
     Find out what the actual workdir root is.
 
-    If ``TMT_WORKDIR_ROOT`` variable is set, it is used as the workdir root.
-    Otherwise, the default of :py:data:`WORKDIR_ROOT` is used.
+    If the ``workdir-root`` cli option is set, it is used as the workdir root.
+    Otherwise, the ``TMT_WORKDIR_ROOT`` environment variable is used if set.
+    If neither is specified, the default value of :py:data:``WORKDIR_ROOT`` is used.
     """
+    if workdir_root_option:
+        return Path(workdir_root_option)
```
Comment (happz) [inline tmt/utils/__init__.py:242]: `workdir_root_option` is already a `Path` instance.

File: tmt/options.py (hunk 9 above)
Comment (happz) [reply tmt/options.py:190]: And the help text? There is no reason to remove mention of default value from help, is it?

File: tmt/cli.py [hunk 12]
```diff
@@ -2043,17 +2050,18 @@ def clean_guests(
     if last and bool(id_):
         raise tmt.utils.GeneralError(
             "Options --last and --id cannot be used together.")
-    if not Path(workdir_root).exists():
+    if workdir_root and not Path(workdir_root).exists():
         raise tmt.utils.GeneralError(f"Path '{workdir_root}' doesn't exist.")
 
     assert context.obj.clean_logger is not None  # narrow type
-
+    workdir_root_path = Path(workdir_root) if workdir_root else None
```
Comment (happz) [inline tmt/cli.py:2057]: This expression can really live in `tmt.Clean()` call below, `workdir_root_path` has no other purpose in life than to be passed to `Clean` class. `workdir_root=effective_workdir_root(Path(workdir_root) if workdir_root else None))` is perfectly fine.

File: tmt/base.py (hunk 6 above)
Comment (happz) [reply tmt/base.py]: Yes, if you start logging it and therefore add a second use of that value, it looks tempting to store it in a local variable. I for one would still use `self.workdir_root`, because it's short, it doesn't harm readability, and brings no performance penalty :)

File: tmt/cli.py (hunk 12 above)
Comment (happz) [reply tmt/cli.py:2057]: Ha, nevermind. How about doing it slightly differently? `workdir_root` and `workdir_root_path` - and also `root_path` above, in `root_path = effective_workdir_root(workdir_root_path)`, the naming is a bit fractured. We need to invent new names because `workdir_root` holds `Optional[str]`, but there's `effective_workdir_root()`, `--workdir-root`, `WORKDIR_ROOT`, so suddenly using `root_path` is just puzzling and taking a turn to new namespace.

Buw, we can ask `option()` to use different name, IIRC, doing this: `option('--workdir-root', '_workdir_root', ...)`. Commands accepting this option then need to change their `workdir_root` parameter to `_workdir_root`, and then we can easily convert `_workdir_root` to a `Path` instance and be done with this mess:

```python
...
@workdir_root_options
@verbosity_options
@dry_options
def clean(context: Context,
          last: bool,
          id_: tuple[str, ...],
          skip: list[str],
          _workdir_root: Optional[str],
          **kwargs: Any) -> None:
    ...
    workdir_root = Path(_workdir_root) if _workdir_root is not None else None

    if workdir_root and not workdir_root.exists():
        raise tmt.utils.GeneralError(f"Path '{workdir_root}' doesn't exist.")

    ...

    if context.invoked_subcommand is None:
        assert context.obj.clean_logger is not None  # narrow type
        root_path = effective_workdir_root(workdir_root)
        ...
```

No unexpected names, the conversion happens as early as possible, workdir root is in `workdir_root`, not in `root_dir` or `workdir_root_path`, and the world will be a bit cleaner. WDYT?

**Comment by skycastlelily** [comment]
So brilliant 🤩,and updated^^

File: tmt/utils/__init__.py (hunk 5 above)
Comment (martinhoyer) [reply tmt/utils/__init__.py]: Sorry I missed this comment. The double to single quotes were not intentional change. I'm just suggesting the wording changes.

**Comment by skycastlelily** [comment]
Yeah,I know, so I updated with  double quotes:)

File: docs/releases.rst [hunk 13]
```diff
@@ -7,6 +7,9 @@
 tmt-1.37.0
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+Add support ``--workdir-root`` for clean function, users can specify
```
Comment (KwisatzHaderach) [inline docs/releases.rst:10]: this needs to be under 1.38 now

**Review by therazix — APPROVED** [review]

File: tmt/base.py (hunk 6 above)
Comment (happz) [reply tmt/base.py]: This does not seem to be addressed yet.

**Comment by skycastlelily** [comment]
ah,yes, I misunderstood you by somehow overlooking "I for one would still
use self.workdir_root" 😅,anyway,rebased and updated^^

File: tmt/cli.py [hunk 14]
```diff
@@ -1930,25 +1935,24 @@ def clean(context: Context,
     exit_code = 0
     if context.invoked_subcommand is None:
         assert context.obj.clean_logger is not None  # narrow type
-
-        # Set path to default
-        context.params['workdir_root'] = tmt.utils.WORKDIR_ROOT
+        root_path = effective_workdir_root(workdir_root)
```
Comment (happz) [inline tmt/cli.py:1938]: One more `root_path` we should root out and replace with `workdir_root`. We can definitely re-use the `workdir_root` name to hold the workdir root, the one we get from `effective_workdir_root`, it should be the same type and meaning.

**Comment by skycastlelily** [comment]
Updated^^

**Comment by happz** [comment]
@skycastlelily please, check the tests, there seems to be some issue with a missing argument. No idea why `pre-commit` and linters did not report it :/

**Comment by skycastlelily** [comment]
Updated,should be good now:)

File: docs/releases.rst [hunk 15]
```diff
@@ -4,10 +4,12 @@
     Releases
 ======================
 
-
```
Comment (happz) [inline docs/releases.rst:7]: This line should not be removed.

**Comment by skycastlelily** [comment]
Updated😅

**Review by happz — APPROVED** [review]

**Review by psss — APPROVED** [review]
Looks good, just one nitpick. Also proposing a minore release note enhancement in 94b9098a. @skycastlelily, does it look goot to you?

**Comment by psss** [comment]
@skycastlelily, what about extending `/tests/clean/basic` a bit to add at least a very basic coverage for the new option?

**Comment by skycastlelily** [comment]
> does it look good to you?

Yeah:)

> Is the new line really needed here?

 I guess so, we need pass workdir_root to Status class,or, let's say,Common
class, so that we can get  self.workdir_root there:)

> what about extending /tests/clean/basic a bit to add at least a very
basic coverage for the new option?

 Done,and updated:)

File: tests/clean/basic/test.sh [hunk 16]
```diff
@@ -4,6 +4,7 @@
 rlJournalStart
     rlPhaseStartSetup
         rlRun "root=/var/tmp/tmt"
+        rlRun "test_root=/var/tmp/test_root"
```
Comment (happz) [inline tests/clean/basic/test.sh:7]: It might be better to use a temporary directory, e.g. like here: https://github.com/teemtee/tmt/blob/main/tests/run/error/test.sh#L8

File: tmt/cli.py [hunk 17]
```diff
@@ -1866,12 +1866,16 @@ def status(
...
+
+    workdir_root = Path(_workdir_root) if _workdir_root is not None else None
+    if workdir_root and not workdir_root.exists():
         raise tmt.utils.GeneralError(f"Path '{workdir_root}' doesn't exist.")
 
     status_obj = tmt.Status(
-        logger=context.obj.logger.clone().apply_verbosity_options(**kwargs),
-        cli_invocation=CliInvocation.from_context(context))
+        logger=context.obj.logger.clone().
+        apply_verbosity_options(**kwargs),
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/cli.py:1876]: >> Is the new line really needed here?
> I guess so, we need pass workdir_root to Status class,or, let's say,Common
class, so that we can get  self.workdir_root there:)

@skycastlelily the question is about the new-line in `logger` argument. Would this be better?

```
        logger=context.obj.logger.clone().apply_verbosity_options(**kwargs),
```

AFAICT the first two parameters to `tmt.Status` did not really change, right? `logger=context.obj.logger.clone().apply_verbosity_options(**kwargs)` and `cli_invocation=CliInvocation.from_context(context)` lines should not change by this patch, except the trailing `)` would be replaced with `,`. This is how it would look:

```python
    status_obj = tmt.Status(
        logger=context.obj.logger.clone().apply_verbosity_options(**kwargs),
        cli_invocation=CliInvocation.from_context(context),
        workdir_root=effective_workdir_root(workdir_root))
```

File: tests/clean/basic/test.sh (hunk 16 above)
Comment (skycastlelily) [reply tests/clean/basic/test.sh]: yeah,updated^^

File: tmt/cli.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (skycastlelily) [reply tmt/cli.py]: Ah,got it now ,thanks for mentor's explain^^

File: tests/clean/basic/test.sh [hunk 18]
```diff
@@ -14,6 +15,13 @@ rlJournalStart
         done
     rlPhaseEnd
 
+    rlPhaseStartTest "Create a couple of runs in non-default root workdir"
+        for id in {001..003}; do
+            rlRun "tmt --feeling-safe run --id clean-$id --workdir_root $test_root"
```
Comment (martinhoyer) [inline tests/clean/basic/test.sh:20]: Afaik, we want to have feeling safe handled by an env var.

File: tmt/cli.py [hunk 19]
```diff
@@ -2053,17 +2062,18 @@ def clean_guests(
     if last and bool(id_):
         raise tmt.utils.GeneralError(
             "Options --last and --id cannot be used together.")
-    if not Path(workdir_root).exists():
+    workdir_root = Path(_workdir_root) if _workdir_root is not None else None
```
Comment (martinhoyer) [inline tmt/cli.py:2065]: I'm not sure if I understand it correctly, but do we need the temporary `_workdir_root`?   
I mean, shouldn't just `if workdir_root and not Path(workdir_root).exists():` be enough?

Comment (happz) [reply tmt/cli.py:2065]: It's on purpose. The idea is, the input is a string, but we want to convert it to `Path` as soon as possible, and use `Path` from that moment on. And `workdir_root` is a natural name for labeling this `Path`. Therefore we ask Click to store the string into `_workdir_root`, then we instantiate the `Path` instance, and assign it to `workdir_root`. If we had `workdir_root` only, a string, then we would need to think about converting it to `Path` when passing it on, e.g. to `tmt.Clean` below. This way, we convert the value to the proper internal type on the boundary.

File: tests/clean/basic/test.sh (hunk 18 above)
Comment (skycastlelily) [reply tests/clean/basic/test.sh]: Updated:)

File: tmt/cli.py (hunk 19 above)
Comment (skycastlelily) [reply tmt/cli.py:2065]: Thanks for your reply ,mentor:)

**Comment by martinhoyer** [comment]
@psss with the test added, perhaps this can get the ready-to-merge label back?

**Comment by skycastlelily** [comment]
:partying_face: :balloon:

**Review by psss — CHANGES_REQUESTED** [review]
Looks good, test needs some adjustments though. I've also moved the release note to the proper section and adjusted the wording a bit.

File: tests/clean/basic/test.sh [hunk 20]
```diff
@@ -14,6 +15,13 @@ rlJournalStart
         done
     rlPhaseEnd
 
+    rlPhaseStartTest "Create a couple of runs in non-default root workdir"
+        for id in {001..003}; do
+            rlRun "tmt run --id clean-$id --workdir_root $test_root"
```
[the remark this answers is not in the document]
Comment (skycastlelily) [reply tests/clean/basic/test.sh]: Yeah, a little awkward :sweat_smile: ,I thought it is just simple copy&paste work, will keep that on mind.Anyway,updated:https://github.com/teemtee/tmt/pull/2850/commits/60409b2920667bb3d25ec56548631c00af7ade1c ,should be good now:)

File: tests/clean/basic/test.sh [hunk 21]
```diff
@@ -38,6 +46,14 @@ rlJournalStart
         rlAssertNotExists "$root/clean-004"
     rlPhaseEnd
 
+    rlPhaseStartTest "Remove plan in a non-default root wordkir"
+        rlRun "tmt clean -v --id clean-002 --workdir_root $test_root"
```
[the remark this answers is not in the document]
Comment (skycastlelily) [reply tests/clean/basic/test.sh]: https://github.com/teemtee/tmt/pull/2850/commits/60409b2920667bb3d25ec56548631c00af7ade1c

**Comment by psss** [comment]
> I've also moved the release note to the proper section and adjusted the wording a bit.

Seems the release note adjustment has been lost during rebase,
added back in 7a0adc40. Please fetch the latest content from te
branch before force-pushing new changes.

There is a couple of tests failing:

    /tests/clean/guests
    /tests/clean/runs
    /tests/status/base

The error seems to be related:

    status() missing 1 required positional argument: 'workdir_root'

Please, have a look. Thanks. Not ready for 1.39, moving to 1.40.

**Comment by skycastlelily** [comment]
> The error seems to be related:
I guess this is because [my commit ](https://github.com/teemtee/tmt/commit/c15154b9edac0240d05da8a3f29f567c5feffb27#diff-2bcd3056539a32d0370d652cfc6c04bd19b674cae84c341c3be7fc51cc3c60e6R1870)to address“ status() missing 1 required positional argument error“ is somehow removed during the rebase，anyway,gonna to have it back:)

**Comment by skycastlelily** [comment]
> I guess it is because [my commit ](https://github.com/teemtee/tmt/commit/c15154b9edac0240d05da8a3f29f567c5feffb27#diff-2bcd3056539a32d0370d652cfc6c04bd19b674cae84c341c3be7fc51cc3c60e6R1870)to address that error is somehow removed

Yeah, that's the problem, added it back:https://github.com/teemtee/tmt/pull/2850/commits/4b61e188446e821bc5c244ee3d2bcef7a11b4dae.  

>  Not ready for 1.39, moving to 1.40.

We have lots of reviews, and commits for this mr, the "new" error is due to the commit to address the error is gone during rebase,hmm,lead, would you please,or do you see any chance we could have this mr released with 1.39 to avoid  rebase mess :crying_cat_face: :pleading_face: ?  :grin:

File: tmt/cli/_root.py [hunk 22]
```diff
@@ -1620,6 +1620,62 @@ def _construct_trying_provision_options(params: Any) -> dict[str, Any]:
...
+    '--abandoned', is_flag=True, default=False,
+    help='List runs which have provision step completed but finish step not yet done.')
+@option(
+    '--active', is_flag=True, default=False,
+    help='List runs where at least one of the enabled steps has not been finished.')
+@option(
+    '--finished', is_flag=True, default=False,
+    help='List all runs which have all enabled steps completed.')
+@verbosity_options
+def status(
```
Comment (psss) [inline tmt/cli/_root.py:1643]: The status definition has been moved to a separate file `status.py`. Please, move the changes there.

**Review by psss — CHANGES_REQUESTED** [review]

File: tmt/cli/_root.py (hunk 22 above)
Comment (skycastlelily) [reply tmt/cli/_root.py]: > The status definition has been moved to a separate file status.py

Yeah,that's why I updated in status.py to fix the error, but I forgot to check the whole change,so failing to note we still have status in _root.py :sweat_smile:  Anyway,addressed:https://github.com/teemtee/tmt/pull/2850/commits/abe8599734da0cba6c7c5a8c6654bc0311399ebd

Comment (psss) [reply tmt/cli/_root.py]: Great, thanks!

**Review by psss — APPROVED** [review]
Now looks good, thanks!

**Comment by skycastlelily** [comment]
> Now looks good, thanks!

Pleasure,I guess we could finally have this one merged? :slightly_smiling_face:

**Comment by psss** [comment]
Yes, just rebased on the latest `main` to get rid of those irrelevant test failures.

**Comment by psss** [comment]
Red tests are irrelevant failures, merging.

**Comment by skycastlelily** [comment]
🎉
